### PR TITLE
Parametric_tw_csv example

### PR DESCRIPTION
**Pull request by dreamer2368**
## Miscellaneous addition to pylibROM
- Bindings for `utils::CSVDatabase` and `utils::HDFDatabase`
- `python_utils/cpp_utils.hpp` now provides a template function `get1DArrayFromPtr` for numpy array.
  - This resolves the issue #11 , where the view numpy array is returned instead of the actual `std::vector` output variable.
  - The example is in line 110 of `utils/pyCSVDatabase.cpp`:
```
    csvdb.def("getIntegerVector", [](
        CSVDatabase &self, const std::string& key, bool append)
    {
        std::vector<int> *datavec = new std::vector<int>;
        self.getIntegerVector(key, *datavec, append);
        return get1DArrayFromPtr(datavec->data(), datavec->size(), true);
    },
```

## `examples/dmd/parametric_tw_csv.py` example demonstration
`parametric_tw_csv.py` demonstrates the parametric DMD, along with the use of CSV/HDF databases.
1. FOM sampling process
  - csv format
```
bash ./heat_conduction_csv.sh
```
  - hdf5 format
```
bash ./heat_conduction_hdf.sh
```
2. Case1: serial DMD
Add `-hdf` tag if you're following the workflow for hdf5.
```
python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -offline
python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -online
``` 
3. Case2: time-windowing DMD
Add `-hdf` tag if you're following the workflow for hdf5.
```
python3 parametric_tw_csv.py -o hc_parametric_tw -nwinsamp 25 -dtc 0.01 -offline
python3 parametric_tw_csv.py -o hc_parametric_tw -nwinsamp 25 -dtc 0.01 -offline
```

### Demonstration results
1. serial DMD, csv
```
Norm of true solution at t = 0.500

000 is 84.539879
Absolute error of DMD solution at t = 0.500000 is 0.106531
Relative error of DMD solution at t = 0.500000 is 0.001260

Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 2.193182e-01 second
Total elapsed time for predicting DMD: 1.743942e+00 second
Average elapsed time for predicting DMD: 3.419494e-02 second
```
little difference from c++ libROM:
```
Norm of true solution at t = 0.5000000000000002 is 84.53987904150704
Absolute error of DMD solution at t = 0.5000000000000002 is 0.1065308142113612
Relative error of DMD solution at t = 0.5000000000000002 is 0.001260124989758469
Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 2.209829e-01 second
Total elapsed time for predicting DMD: 1.791613e+00 second
Average elapsed time for predicting DMD: 3.512967e-02 second
```
2. serial DMD, hdf
```
Norm of true solution at t = 0.500000 is 84.539879
Absolute error of DMD solution at t = 0.500000 is 0.106531
Relative error of DMD solution at t = 0.500000 is 0.001260

Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 2.122204e-01 second
Total elapsed time for predicting DMD: 1.759935e+00 second
Average elapsed time for predicting DMD: 3.450852e-02 second
```
similar result from c++ libROM:
```
Norm of true solution at t = 0.5000000000000002 is 84.53987904150704
Absolute error of DMD solution at t = 0.5000000000000002 is 0.1065308142113612
Relative error of DMD solution at t = 0.5000000000000002 is 0.001260124989758469
Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 1.646451e-01 second
Total elapsed time for predicting DMD: 1.757581e+00 second
Average elapsed time for predicting DMD: 3.446237e-02 second
```
3. time-windowing, csv
```
Norm of true solution at t = 0.500000 is 84.539879
Absolute error of DMD solution at t = 0.500000 is 0.055013
Relative error of DMD solution at t = 0.500000 is 0.000651

Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 1.072407e-01 second
Total elapsed time for predicting DMD: 2.882011e-01 second
Average elapsed time for predicting DMD: 5.651002e-03 second
```
c++ libROM:
```
Norm of true solution at t = 0.5000000000000002 is 84.53987904150704
Absolute error of DMD solution at t = 0.5000000000000002 is 0.0550131373888892
Relative error of DMD solution at t = 0.5000000000000002 is 0.0006507359368456049
Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 8.679060e-02 second
Total elapsed time for predicting DMD: 2.865902e-01 second
Average elapsed time for predicting DMD: 5.619415e-03 second
```
4. time-windowing, hdf
```
Norm of true solution at t = 0.500000 is 84.539879
Absolute error of DMD solution at t = 0.500000 is 0.055013
Relative error of DMD solution at t = 0.500000 is 0.000651

Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 1.108568e-01 second
Total elapsed time for predicting DMD: 3.017945e-01 second
Average elapsed time for predicting DMD: 5.917540e-03 second
```
c++ libROM:
```
Norm of true solution at t = 0.5000000000000002 is 84.53987904150704
Absolute error of DMD solution at t = 0.5000000000000002 is 0.05501313738888904
Relative error of DMD solution at t = 0.5000000000000002 is 0.000650735936845603
Elapsed time for training DMD: 0.000000e+00 second
Elapsed time for preprocessing DMD: 7.375104e-02 second
Total elapsed time for predicting DMD: 2.931510e-01 second
Average elapsed time for predicting DMD: 5.748059e-03 second
```

### CODE REVIEW

**Review by bumpwy — APPROVED** [review]

File: examples/dmd/parametric_tw_csv.py [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+# // Parametric serial DMD command (for HDF version, append -hdf):
+# //   python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -offline
+# //   python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -online
+# //
+# // Final-time prediction error (Last line in run/hc_parametric_serial/dmd_par5_prediction_error.csv):
+# //   0.0012598331433506
+# //
+# // Parametric time windowing DMD command (for HDF version, append -hdf):
+# //   python3 parametric_tw_csv.py -o hc_parametric_tw -nwinsamp 25 -dtc 0.01 -offline
+# //   python3 parametric_tw_csv.py -o hc_parametric_tw -nwinsamp 25 -dtc 0.01 -offline
```
Comment (chldkdtn) [inline examples/dmd/parametric_tw_csv.py]: Do you want it to be -online?

File: examples/dmd/parametric_tw_csv.py [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+# // Compile with: make parametric_tw_csv
+# //
+# // Generate CSV or HDF database on heat conduction with either
+# // heat_conduction_csv.sh or heat_conduction_hdf.sh (HDF is more efficient).
+# //
+# // =============================================================================
+# //
+# // Parametric serial DMD command (for HDF version, append -hdf):
+# //   python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -offline
+# //   python3 parametric_tw_csv.py -o hc_parametric_serial -rdim 16 -dtc 0.01 -online
```
Comment (chldkdtn) [inline examples/dmd/parametric_tw_csv.py:19]: Without -list directory specified, how does it know which parameter to test?

Comment (dreamer2368) [reply examples/dmd/parametric_tw_csv.py:19]: I don't exactly know the full workflow, but it does have `-list` option, which is set to `dmd_list` by default. `heat_conduction_csv.sh` (or hdf) uses the default setting, so the command does not need to specify `-list`.

File: examples/dmd/parametric_tw_csv.py (hunk 1 above)
Comment (dreamer2368) [reply examples/dmd/parametric_tw_csv.py]: This is fixed now.

**Review by chldkdtn — APPROVED** [review]

**Pull request merged**